File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    helper(current,max){
        if(!current) return 0
        let left=this.helper(current.left,max)
        let right=this.helper(current.right,max)
        let sum= current.val+Math.max(0,left)+Math.max(0,right)
        max[0]=Math.max(max[0],sum)
        return Math.max(0,left,right)+current.val
    }
    maxPathSum(root) {
        let max=[-1001]
        this.helper(root,max)
        return max[0]
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
      if not root:
        return 0
      diameter=[0]
      self.helper(root,diameter)
      return diameter[0]
    def helper(self,cur,diameter):
      if not cur:
        return 0
      left,right=self.helper(cur.left,diameter),self.helper(cur.right,diameter)
      diameter[0]=max(diameter[0],left+right)
      return 1+max(left,right)
      
